File: microservices/shared/src/events/base.event.ts
import { v4 as uuidv4 } from 'uuid';

export interface BaseEvent {
  eventId: string;
  eventType: string;
  tenantId: string;
  timestamp: string;
  version: string;
  correlationId?: string;
  causationId?: string;
  metadata?: Record<string, unknown>;
}

export function createEvent<T extends Omit<BaseEvent, 'eventId' | 'timestamp' | 'version'>>(
  event: T
): T & { eventId: string; timestamp: string; version: string } {
  return {
    ...event,
    eventId: uuidv4(),
    timestamp: new Date().toISOString(),
    version: '1.0',
  };
}

File: common/src/main/java/com/teasui/common/event/BaseEvent.java
package com.teasui.common.event;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;

import java.time.Instant;
import java.util.UUID;

@Data
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "eventType")
@JsonSubTypes({
        @JsonSubTypes.Type(value = TenantEvent.class, name = "TENANT"),
        @JsonSubTypes.Type(value = WorkflowEvent.class, name = "WORKFLOW"),
        @JsonSubTypes.Type(value = NotificationEvent.class, name = "NOTIFICATION")
})
public abstract class BaseEvent {

    private String eventId;
    private String eventType;
    private String tenantId;
    private Instant occurredAt;
    private String correlationId;
    private int version;

    public static String generateEventId() {
        return UUID.randomUUID().toString();
    }
}
